algorithms/src/std_algorithms/Kokkos_ModifyingOperations.hpp

```diff
@@ -0,0 +1,118 @@
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_MODIFYING_OPERATIONS_HPP
+#define KOKKOS_MODIFYING_OPERATIONS_HPP
+
+#include <Kokkos_Core.hpp>
+#include "Kokkos_BeginEnd.hpp"
+#include "Kokkos_Constraints.hpp"
+
+/// \file Kokkos_ModifyingOperations.hpp
+/// \brief Kokkos modifying operations
+
+namespace Kokkos {
+namespace Experimental {
+
+// see https://github.com/kokkos/kokkos/issues/4075
+//
+// move
+// swap
+// iter_swap
+
+// move
+template <typename T>
+KOKKOS_INLINE_FUNCTION std::remove_reference_t<T>&& move(T&& t) {
+  return static_cast<std::remove_reference_t<T>&&>(t);
+}
+
+// swap
+template <class T>
+KOKKOS_INLINE_FUNCTION void swap(T& a, T& b) noexcept {
+  static_assert(
+      std::is_move_assignable<T>::value && std::is_move_constructible<T>::value,
+      "Kokkos::Experimental::swap arguments must be move assignable "
+      "and move constructible");
+
+  T tmp = ::Kokkos::Experimental::move(a);
+  a     = ::Kokkos::Experimental::move(b);
+  b     = ::Kokkos::Experimental::move(tmp);
+}
+
+//----------------------------------------------------------------------------
+namespace Impl {
+template <class IteratorType1, class IteratorType2>
+struct StdIterSwapFunctor {
+  IteratorType1 m_a;
+  IteratorType2 m_b;
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(int i) const {
+    (void)i;
+    ::Kokkos::Experimental::swap(*m_a, *m_b);
+  }
+
+  KOKKOS_INLINE_FUNCTION
+  StdIterSwapFunctor(IteratorType1 _a, IteratorType2 _b) : m_a(_a), m_b(_b) {}
+};
+
+template <class IteratorType1, class IteratorType2>
+void iter_swap_impl(IteratorType1 a, IteratorType2 b) {
+  Kokkos::parallel_for(1,
+                       StdIterSwapFunctor<IteratorType1, IteratorType2>(a, b));
+  Kokkos::fence();
```

Comment:
Once you settle on a set of necessary fences, would you mind adding string labels to them?

`Kokkos::fence("iter_swap: fence after operation");`

Definitely wait til you have the right set of fences, though

Reply:
yes definitely!

Reply:
Ok, I might be missing something obvious, but I couldn't find a `fence()` overload accepting labels.

Compiler output:
```
In file included from ../algorithms/unit_tests/../src/std_algorithms/Kokkos_NonModifyingSequenceOperations.hpp:51:
../algorithms/unit_tests/../src/std_algorithms/Kokkos_ModifyingOperations.hpp:104:3: error: too many arguments to function call, expected 0, have 1; did you mean 'Kokkos_Tools_ToolProgrammingInterface::fence'?
  ::Kokkos::fence("label");
  ^~
  Kokkos_Tools_ToolProgrammingInterface::fence
../core/src/impl/Kokkos_Profiling_C_Interface.h:140:41: note: 'Kokkos_Tools_ToolProgrammingInterface::fence' declared here
  Kokkos_Tools_toolInvokedFenceFunction fence;
                                        ^
1 error generated.
```

Reply:
When did you rebase last?
```
$ git grep -n " fence(const std::string&"
core/src/Cuda/Kokkos_Cuda_Instance.hpp:156:  void fence(const std::string&) const;
core/src/Kokkos_Core.hpp:190:void fence(const std::string&);
core/src/Kokkos_Cuda.hpp:190:  void fence(const std::string&) const;
core/src/Kokkos_Cuda.hpp:276:  void fence(const std::string&) final;
core/src/Kokkos_HIP_Space.hpp:589:  void fence(const std::string&) const;
core/src/Kokkos_HIP_Space.hpp:650:  void fence(const std::string&) final;
core/src/Kokkos_OpenMP.hpp:111:  void fence(const std::string& name) const;
core/src/Kokkos_OpenMP.hpp:192:  void fence(const std::string&) final;
core/src/Kokkos_OpenMPTarget.hpp:94:  static void fence(const std::string&);
core/src/Kokkos_OpenMPTarget.hpp:146:  void fence(const std::string&) final;
core/src/Kokkos_SYCL.hpp:117:  void fence(const std::string&) const;
core/src/Kokkos_SYCL.hpp:172:  void fence(const std::string&) final;
core/src/Kokkos_Serial.hpp:167:  void fence(const std::string& name) const {
core/src/Kokkos_Serial.hpp:232:  void fence(const std::string&) final;
core/src/Kokkos_Threads.hpp:114:  void fence(const std::string&) const;
core/src/Kokkos_Threads.hpp:197:  void fence(const std::string&) final;
core/src/OpenMPTarget/Kokkos_OpenMPTarget_Instance.hpp:64:  void fence(const std::string& name,
core/src/impl/Kokkos_Core.cpp:1098:void fence(const std::string& name) { Impl::fence_internal(name); }
```

Reply:
Exactly that, I forgot to rebase. Thanks for the tip!
Anyways - labels are added, I went with the description that David has suggested.